### PR TITLE
Playwright: Migrate the Plans specs and POMs

### LINKED ISSUE #54842: Playwright: Migrate the wp-plan specs
### Details

Clean-up and migrate the `wp-plan` specs and migrate them and their POM to Playwright

### Checklist

- [ ] wp-plan-purchase_comparing-spec.js
- [ ] wp-plan-purchase_renew-spec.js
- [ ] wp-plan-purchase_upgrade-spec.js
- [ ] wp-plan-purchase_viewing-spec.js

### Related

_No response_

### PR DESCRIPTION
**Pull request by dpasque**
#### Changes proposed in this Pull Request

Playwright implementation of the old Plans specs in the `specs-calypso` directory.

Specifically, it migrates, with some modification, the coverage from `specs/specs-calypso`...
- `wp-plan-purchase__comparing-spec.js`
- `wp-plan-purchase__renew-spec.js`
- `wp-plan-purchase__upgrade-spec.js`

It does **_not_** include coverage from `wp-plan-purchase__viewing-spec.js`, which is actually a misnomer because it really just covers applying a coupon in the cart. Coupon coverage should be done in a separate, more isolated test case.

These POMs take the new design approach of not using a `BaseContainer` parent class for the POM classes (as outlined in PR #54697 - option A), and instead bakes the needed waiting into class methods.

#### Testing instructions

Run the Playwright tests, and make sure they pass!

Fixes #54842
Implements option A outlined in PR #54697

### CODE REVIEW

File: packages/calypso-e2e/src/lib/pages/plans-page.ts
Comment (dpasque) [inline packages/calypso-e2e/src/lib/pages/plans-page.ts]: I can't decide if I want to do these this way, where we do a text-based selector, and let `waitForSelector` throw, or alternatively grab and return the actual content of the text nodes at these locations, then have the tests themselves assert (using Jest `expect`) against that string to make sure it is the expected value.

File: packages/calypso-e2e/src/lib/pages/individual-purchase-page.ts
Comment (dpasque) [inline packages/calypso-e2e/src/lib/pages/individual-purchase-page.ts]: I went back and forth a lot on this design pattern used throughout this PR. The way I see it, there are two different ways we could do this:
1. Find and return the string of whatever is at this text node, then `expect` against it in the `spec` file
2. Do a `waitForSelector` using the expected text as part of the selector.

I opted for #2 because it seemed to give the most stability to the test. Say, for example, that it takes a half second for the text on a page to update. We wouldn't want that test to fail - we would want to rely on auto-waiting for that update to occur.

File: packages/calypso-e2e/src/lib/pages/individual-purchase-page.ts
Comment (dpasque) [inline packages/calypso-e2e/src/lib/pages/individual-purchase-page.ts]: This test plan required selectors that might need to take in different text to include as part of the selector. I decided to opt for splitting the selectors at the top of the file into those that are static vs those that are dynamic so that people don't accidentally include a pointer to a function (in the case of the dynamic selectors) as a selector in a test (instead of invoking that function to get the selector).

**Comment by worldomonation** [comment]
I've triggered the Playwright tasks for mobile and desktop :)

Edit: looks like as part of this change, all other tests that use the Sidebar need to be changed to remove the use of `Expect` as well.

File: packages/calypso-e2e/src/lib/components/sidebar-component.ts
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/components/sidebar-component.ts]: To clarify; I assume we would need to rebase any in-progress branches once this lands on trunk?

File: packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts]: nit: Full stop at end of sentence :)

File: packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts]: nit: JSDoc in calypso should be in format of:
` @param {type} variableName description`

File: packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts:12]: Interesting, I have not seen this before, but definitely useful in other situations I've faced.

Question: is it possible to have it all grouped under `const selectors = { ... }`? Or does it have to be a separate constant?

File: packages/calypso-e2e/src/lib/pages/individual-purchase-page.ts
Comment (worldomonation) [reply packages/calypso-e2e/src/lib/pages/individual-purchase-page.ts]: Interesting, I have not seen this before, but definitely useful in other situations I've faced.

I asked a question originally about 'can we keep the unified `const selectors` instead of splitting', but you've touched on why it should be split. Thanks for explaining.

Stylistically I would prefer to retain the unified `selectors` object and leave the responsibility to the test writer to figure out if it's a static/dynamic selector. I would assume if they accidentally reference a dynamic selector without invoking it the step involved would fail anyway.

But this is not a blocker and personally I think the benefits of dynamic selectors far outweigh any stylistic qualms I may have.

File: packages/calypso-e2e/src/lib/pages/login-page.ts
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/pages/login-page.ts:61]: I think I also changed something in the LoginFlow in #54848:
```
await Promise.all( [ this.page.waitForNavigation(), this.baseflow() ] );
```
Perhaps having both isn't necessary, so I'd leave it up to you as to which one you'd want to keep.

File: packages/calypso-e2e/src/lib/pages/plans-page.ts
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/pages/plans-page.ts]: Just a thought - since the desktop and mobile version only differ by the first bit (`.plan-features__<viewport>`, would this selector work instead?

```
.plan-features__${viewport} >> .plan-features__actions-button.is-${ plan.toLowerCase() }-plan:has-text("${ buttonText }")
```

By unifying the selector it should be possible to call the same selector, just with an additional param for viewport:

```
const viewportType = getViewportName();
const selector = dynamicSelectors.mobileActionButton( { viewport: viewport, plan: plan, buttonText: buttonText } )
```

If you have already tried unifying the two selectors and it didn't work, then sorry about that!

File: test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts
Comment (worldomonation) [inline test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts]: Personally, I would prefer to explicitly perform actions in POMs, even if the action may not be necessary (like in this case), for a number of reasons:

- guards against situation where default landing tab in Upgrades >> Plans change from 'My Plan' to something else in the future.
- the e2e spec reads more like  series of commands that _should work_

Would you be open to something like this?

```
// in PlansPage

async clickTab( name: PlansPageTab ) {
    // click on tab
    // check that tab is indeed clicked
}
```

An example of this can be found in [MarketingPage](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-e2e/src/lib/pages/marketing-page.ts#L28) and [MediaPage](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-e2e/src/lib/pages/media-page.ts#L86).

My opinion is that generally speaking, having more `action` oriented methods that check whether the action was successful) is preferable. `validation` oriented methods have their place, but should serve as helpers in situations where the action method could not verify for some reason, or validation needs to happen independently of action (like in this case, as demonstrated by the steps below where Upgrades >> Plans - Plans tab is automatically landed on when cart item is removed).

File: test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts
Comment (worldomonation) [inline test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts]: Nice!

File: test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts
Comment (worldomonation) [inline test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts]: Good explanation, thanks!

File: test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts
Comment (worldomonation) [inline test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts]: Could you add in to the title of the suite here about upgrading to WordPress.com Business?

File: test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts
Comment (worldomonation) [inline test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts]: Could you specify what plan is active here currently? (WordPress.com Premium), just for clarify when reading the output.

File: packages/calypso-e2e/src/lib/components/sidebar-component.ts
Comment (dpasque) [reply packages/calypso-e2e/src/lib/components/sidebar-component.ts]: Yep! Anything that is calling `SidebarComponent.Expect` would need to be updated to just use the constructor. But everything after that should work just the same.

File: packages/calypso-e2e/src/lib/pages/login-page.ts
Comment (dpasque) [reply packages/calypso-e2e/src/lib/pages/login-page.ts:61]: Ahh, good catch! I think I'll keep this in the LoginPage for now as it keeps the navigation and page load waits closer to the action that triggers them, and can be used outside the flow if needed! Let me know if I should switch it back the other way!

File: test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts
Comment (dpasque) [reply test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts]: Sure, that makes a lot of sense! I updated the start of the test to reflect that, and left it in for the test for the automatic navigation after cart removal.

File: packages/calypso-e2e/src/lib/pages/individual-purchase-page.ts
Comment (dpasque) [reply packages/calypso-e2e/src/lib/pages/individual-purchase-page.ts]: That makes sense, I updated it to be in the same object

File: packages/calypso-e2e/src/lib/pages/login-page.ts
Comment (worldomonation) [reply packages/calypso-e2e/src/lib/pages/login-page.ts:61]: Sounds good. I don't think the additional layer of `waitForNavigation()` will trip anything up, but maybe it should be removed from the `LoginFlow` side of things in a separate PR.

**Comment by dpasque** [comment]
Hrmm, getting some failures in other playwright specs, I'll look into it and see if there's a missing wait I've now accidentally taken out in the refactors.

**Comment by dpasque** [comment]
> Hrmm, getting some failures in other playwright specs, I'll look into it and see if there's a missing wait I've now accidentally taken out in the refactors.

Turns out this was an artifact of taking out the sidebar wait in the new post flow, which i have restored.

This has me thinking: there seems to be a coupling between the navbar and the sidebar. The navbar is the first to load, but it will basically just swallow all clicks until the javascript that renders the sidebar is done rendering.

This kind of makes me wonder if we should just add an element of sidebar waiting to the navbar component before it takes any of its actions. It would only add milliseconds, and we could of course cache for when folks need to take multiple actions on the same navbar.

What do you think @worldomonation? It feels weird to couple the sidebar to the navbar in the navbar class... But it might make it easier for new test writers to avoid this race condition.

File: packages/calypso-e2e/src/lib/pages/individual-purchase-page.ts
Comment (worldomonation) [reply packages/calypso-e2e/src/lib/pages/individual-purchase-page.ts]: Nice!

File: test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts
Comment (worldomonation) [inline test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts]: Oh, I missed this on the first pass, but one more nit - 

Naming structure of files uses a double underscore immediately after the major feature being tested, so in this case your filename should read:

`wp-plans__purchases-spec.ts`.

File: test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts
Comment (worldomonation) [inline test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts]: I used to do this when we didn't have automatic import sorting, but now that we do, I think the `/** constants */` block can go away.

File: test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts
Comment (worldomonation) [inline test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts]: Just found the name funny 😆 

`plan in plan`

**Comment by worldomonation** [comment]
> > Hrmm, getting some failures in other playwright specs, I'll look into it and see if there's a missing wait I've now accidentally taken out in the refactors.
> 
> Turns out this was an artifact of taking out the sidebar wait in the new post flow, which i have restored.
> 
> This has me thinking: there seems to be a coupling between the navbar and the sidebar. The navbar is the first to load, but it will basically just swallow all clicks until the javascript that renders the sidebar is done rendering.
> 
> This kind of makes me wonder if we should just add an element of sidebar waiting to the navbar component before it takes any of its actions. It would only add milliseconds, and we could of course cache for when folks need to take multiple actions on the same navbar.
> 
> What do you think @worldomonation? It feels weird to couple the sidebar to the navbar in the navbar class... But it might make it easier for new test writers to avoid this race condition.

Yeah, I remember now that was the reason why I had to wait on the Sidebar first. Initially, it was waiting on the main page content to load, but that was shot down in earlier reviews so the best alternative was Sidebar.

With that said though, I think I can take a look at this issue separately from this PR, if you'd like. Now that I know how to manipulate Playwright and TS some more, I think I can come at this a bit better.

Does having to revert your changes to the `NewPostFlow` impact your implementation of POMs using Option A at all? If it does, what we can do is to have a separate PR.

File: test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts
Comment (worldomonation) [inline test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts]: Issue #54858 references this PR - perhaps a typo in the issue number here? Either way, good find!

File: test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts
Comment (dpasque) [reply test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts]: Ha, right? I couldn't come up with anything better 😆  So I'm very open to ideas!

File: test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts
Comment (dpasque) [reply test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts]: Lols, this is what I get for having multiple tabs open haha...

File: test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts
Comment (worldomonation) [inline test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts]: Not a nit or a suggestion to change, just sharing a neat approach I found the other day looking for 'conditional skip' in Jest: [link](https://stackoverflow.com/a/60438234).

Comment (dpasque) [reply test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts]: Oh I love this!! I will do something, maybe like "describeDesktopOnly"

**Comment by dpasque** [comment]
> Does having to revert your changes to the NewPostFlow impact your implementation of POMs using Option A at all? If it does, what we can do is to have a separate PR.

No, not really actually! The way I see it is that no matter what, we have to do some kind of waiting on the sidebar in a way that's re-usable. In this case, I just moved that to an explicit method that could be used in those places. I personally think that reads more cleanly than making assumptions about constructor-like functions.

File: test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts
Comment (worldomonation) [reply test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts]: Since the code is validating that the current plan that is active in the `My Plans` tab, if you want to change it my suggestions would be :

- validateActivePlan
- checkActivePlan

and a couple other variations on the word `check/validate`.

Correct me if I'm wrong, but it seems that `active plan` is only shown within Upgrades >> Plans : Plans tab, so I think it's unnecessary to specify `in my plan`.

Comment (dpasque) [reply test/e2e/specs/specs-playwright/wp-plans_purchases-spec.ts]: >Correct me if I'm wrong, but it seems that active plan is only shown within Upgrades >> Plans : Plans tab, so I think it's unnecessary to specify in my plan.

Well, there is really two ways to validate this: the titular card in the "my plan" tab, and whichever card has "manage" in the "plans" tab, so I wanted some distinction.

It's kind of verbose, but what about "validateActivePlanInMyPlanTab"?

**Review by worldomonation — APPROVED** [review]
Otherwise, this change looks good to me! As long as it continues to pass on both mobile and desktop, I'm happy to see it 🚀

File: packages/calypso-e2e/src/lib/pages/individual-purchase-page.ts
Comment (WunderBart) [inline packages/calypso-e2e/src/lib/pages/individual-purchase-page.ts:4]: Lovin' the dynamic selectors! 🙌

File: packages/calypso-e2e/src/lib/pages/plans-page.ts
Comment (WunderBart) [inline packages/calypso-e2e/src/lib/pages/plans-page.ts:24]: ~Looks like a typo? `table` -> `tablet`~

Comment (WunderBart) [reply packages/calypso-e2e/src/lib/pages/plans-page.ts:24]: Oh, no it's not - just checked! 😅 My bad!

**Pull request merged**
